homeassistant/components/sensor/hue.py

```diff
@@ -0,0 +1,195 @@
+        response = requests.get(self.url)
+        if response.status_code != 200:
+            _LOGGER.warning("Invalid response from API")
+        else:
+            self.data = parse_hue_api_response(response.json())
+
+
+class HueSensor(Entity):
+    """Class to hold Hue Sensor basic info."""
+
+    ICON = 'mdi:run-fast'
+
+    def __init__(self, hue_id, data):
+        """Initialize the sensor object."""
+        self._hue_id = hue_id
+        self._data = data    # data is in .data
+        self._icon = None
+        self._name = self._data.data[self._hue_id]['name']
+        self._model = self._data.data[self._hue_id]['model']
+        self._state = self._data.data[self._hue_id]['state']
+        self._attributes = {}
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def icon(self):
+        """Icon to use in the frontend, if any."""
+        return self._icon
+
+    @property
+    def device_state_attributes(self):
+        """Only motion sensors have attributes currently, but could extend."""
+        return self._attributes
+
+    def update(self):
+        """Update the sensor."""
+        self._data.update()
+        self._state = self._data.data[self._hue_id]['state']
+        if self._model == 'SML001':
+            self._icon = 'mdi:run-fast'
+            self._attributes['light_level'] = self._data.data[
+                self._hue_id]['light_level']
+            self._attributes['temperature'] = self._data.data[
+                self._hue_id]['temperature']
+        elif self._model in ['RWL021', 'ZGPSWITCH']:
+            self._icon = 'mdi:remote'
+            self._attributes['last updated'] = self._data.data[
+                self._hue_id]['last_updated']
+        elif self._model == 'Geofence':
+            self._icon = 'mdi:cellphone'
+
+
+def parse_hue_api_response(response):
+    """Take in the Hue API json response."""
+    data_dict = {}    # The list of sensors, referenced by their hue_id.
+
+    # Loop over all keys (1,2 etc) to identify sensors and get data.
+    for key in response.keys():
+        sensor = response[key]
+
+        if sensor['modelid'] in ['RWL021', 'SML001', 'ZGPSWITCH']:
+            _key = sensor['uniqueid'].split(':')[-1][0:5]
+
+            if sensor['modelid'] == 'RWL021':
+                data_dict[_key] = parse_rwl(sensor)
+            elif sensor['modelid'] == 'ZGPSWITCH':
+                data_dict[_key] = parse_zpg(sensor)
+            else:
+                if _key not in data_dict.keys():
+                    data_dict[_key] = parse_sml(sensor)
+                else:
+                    data_dict[_key].update(parse_sml(sensor))
+
+        elif sensor['modelid'] == 'HA_GEOFENCE':
+            data_dict['Geofence'] = parse_geofence(sensor)
+    return data_dict
+
+
+def parse_sml(response):
```

Comment:
This is too much protocol specific information. Please extract this to a new 3rd party library